mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/loader/MassIndexingTypeLoadingContext.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.massindexing.loader;
+
+import java.util.Map;
+import org.hibernate.search.mapper.pojo.massindexing.loader.dsl.MassIndexingTypeBatchLoader;
+import org.hibernate.search.mapper.pojo.massindexing.loader.dsl.MassIndexingTypeBatchingStep;
+
+/**
+ * The DSL entry point passed to consumers in
+ *
+ * @param <E> The resulting entity type (output)
+ * @param <C> The contextural indexer information
+ */
+public interface MassIndexingTypeLoadingContext<E, C extends MassIndexerContext> {
+
+	/**
+	 * Get the contextural indexer information.
+	 *
+	 * @return contextural indexer information.
+	 */
+	C indexer();
+
+	/**
+	 * @return the identifier of the tenant.
+	 */
+	String tenantId();
+
+	/**
+	 * Get the context data for this indexing query.
+	 *
+	 * @return context data.
+	 */
+	Map<Object, Object> contextData();
+
+	/**
+	 * Set the option for inxexing process.
+	 *
+	 * @param value The param value.
+	 * @return {@code this} for method chaining.
+	 */
+	MassIndexingTypeLoadingContext<E, C> totalCount(long value);
+
+	/**
+	 * Get the in progress indexing process,
+	 *
+	 * @return true if process is active.
+	 */
+	boolean active();
+
+	/**
+	 * Get the entity names indexing process,
+	 *
+	 * @return the entity names by separated comma list.
+	 */
+	String includedEntityNames();
+
+	/**
+	 * Get the entity super type indexing process,
+	 *
+	 * @return the entity super type.
+	 */
+	Object commonSuperType();
```

Comment:
It is required to filter the data source by type. Otherwise, it would always index the entire source, no matter what super type. 

```java
		Set<?> identifiers = source.entrySet().stream()
				.filter( ent -> context.commonSuperType().isAssignableFrom( ent.getValue().getClass() ) )
				.filter( ent -> context.indexed( ent.getValue() ) )
				.map( Entry::getKey ).collect( Collectors.toSet() );
		Iterator<?> iterator = identifiers.iterator();
```
test:
```java
org.hibernate.search.integrationtest.mapper.pojo.massindexing.MassIndexingComplexHierarchyIT#
```

Reply:
Moved to "MassIndexingThreadContext". Necessary for source filtering. According to the test "MassIndexingComplexHierarchyIT", the same as in mapping hibernate, I assume that the source may contain subsequent instances of types.